libdrgn/arch_s390.c

```diff
@@ -0,0 +1,327 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "error.h"
+#include "register_state.h"
+#include "platform.h" // IWYU pragma: associated
+
+/*
+ * The ABI specification can be found at:
+ * https://github.com/IBM/s390x-abi
+ */
+
+#include "arch_s390_defs.inc"
+
+static const struct drgn_cfi_row default_dwarf_cfi_row_s390 = DRGN_CFI_ROW(
+	// The psABI defines that callee-saved registers default to
+	// DW_CFA_same_value.
+	DRGN_CFI_SAME_VALUE_INIT(DRGN_REGISTER_NUMBER(r6)),
+	DRGN_CFI_SAME_VALUE_INIT(DRGN_REGISTER_NUMBER(r7)),
+	DRGN_CFI_SAME_VALUE_INIT(DRGN_REGISTER_NUMBER(r8)),
+	DRGN_CFI_SAME_VALUE_INIT(DRGN_REGISTER_NUMBER(r9)),
+	DRGN_CFI_SAME_VALUE_INIT(DRGN_REGISTER_NUMBER(r10)),
+	DRGN_CFI_SAME_VALUE_INIT(DRGN_REGISTER_NUMBER(r11)),
+	DRGN_CFI_SAME_VALUE_INIT(DRGN_REGISTER_NUMBER(r12)),
+	DRGN_CFI_SAME_VALUE_INIT(DRGN_REGISTER_NUMBER(r13)),
+	DRGN_CFI_SAME_VALUE_INIT(DRGN_REGISTER_NUMBER(r14)),
+	DRGN_CFI_SAME_VALUE_INIT(DRGN_REGISTER_NUMBER(r15)),
+	);
+
+static struct drgn_error drgn_invalid_rel = {
+	.code = DRGN_ERROR_OTHER,
+	.message = "invalid relocation entry",
+};
+
+static struct drgn_error *
+apply_rela_bits(char *dest, uint64_t loc, uint64_t val,
```

Comment:
This seems to make assumptions that the byte order and word size match that of s390x, which may not be true. Please use the existing relocation helper functions that handle byte order and unaligned accesses. We don't need to handle all of these relocation forms, since only a few tend to be used for debug info.